### PR TITLE
Fix for weight loading bug with `channels_first`

### PR DESCRIPTION
**Pull request by the-moliver**
@fchollet This is the start of a fix for the `channels_first` weight loading bug. The bug is due to loading weights from Keras 1 models that used the Tensorflow backend. The current weight loading code assumes all Keras 1 models have Theano dim ordering and so things get messed up. Also the `original_backend` is not found automatically in the old models and so is set to `None`, which makes it difficult or impossible to always convert correctly. Thus I made it so the user can specify the `original_backend` so things can be appropriately converted. Note that I removed `layer_utils.convert_all_kernels_in_model(model)` from the vgg16 example since the weight loading code takes care of the kernel flipping already.

TODO:
1) Properly handle `Conv1D ` and `Conv2DTranspose`. I wasn't sure exactly what to do here so I left it alone, but the same issues probably apply.
2) Fix other applications where weight loading is a problem.

### CODE REVIEW

**Comment by 111Moderato** [comment]
Thank you very much for your help！

File: keras/engine/topology.py [hunk 1]
```diff
@@ -2831,7 +2834,7 @@ def preprocess_weights_for_loading(layer, weights,
                 weights[0] = np.transpose(weights[0], (2, 3, 0, 1))
 
         if layer.__class__.__name__ == 'Conv3D':
-            if layer.data_format == 'channels_first':
+            if layer.data_format == 'channels_first' and original_backend in {'th', 'theano'}:
```
Comment (the-moliver) [inline keras/engine/topology.py]: Not if the old `dim_ordering` was `'tf'`:
https://github.com/fchollet/keras/blob/1.2.2/keras/layers/convolutional.py#L422
Transposing in this case messes things up and is the main source of failure.

Comment (the-moliver) [inline keras/engine/topology.py]: Maybe specifying `original_dim_ordering` would be better, but I was trying to use the arguments present and it seemed like `dim_ordering` and the backend were more coupled in Keras 1 so I thought it would be ok.

File: keras/engine/topology.py [hunk 2]
```diff
@@ -2785,6 +2785,9 @@ def preprocess_weights_for_loading(layer, weights,
         A list of weights values (Numpy arrays).
     """
     if original_keras_version == '1':
+        if original_backend is None:
+            raise ValueError('Please specify original_backend')
```
Comment (the-moliver) [inline keras/engine/topology.py]: Fair enough.

File: keras/engine/topology.py [hunk 3]
```diff
@@ -2815,7 +2818,7 @@ def preprocess_weights_for_loading(layer, weights,
             weights[0] = weights[0][:, 0, :, :]
 
         if layer.__class__.__name__ == 'Conv2D':
-            if layer.data_format == 'channels_first':
+            if layer.data_format == 'channels_first' and original_backend in {'th', 'theano'}:
```
Comment (the-moliver) [inline keras/engine/topology.py]: @fchollet It doesn't seem to know whether the old model had dim_ordering was 'tf' or 'th', and transposing is not right if it was 'tf':
https://github.com/fchollet/keras/blob/1.2.2/keras/layers/convolutional.py#L422
Transposing in this case messes things up and is the main source of failure.

File: keras/engine/topology.py [hunk 4]
```diff
@@ -2504,7 +2504,7 @@ def save_weights(self, filepath, overwrite=True):
         f.flush()
         f.close()
 
-    def load_weights(self, filepath, by_name=False):
+    def load_weights(self, filepath, by_name=False, original_backend=None):
```
Comment (the-moliver) [inline keras/engine/topology.py]: Oh definitely, I first wanted to assess the viability of this approach.

**Comment by fchollet** [comment]
The original backend is orthogonal to the original data format.

- backend information is only used to convert kernel weight values (unrelated to shape) due to differences in convolution implementation (TF does correlation)
- only data_format / dim_ordering affects whether or not we should be transposing the kernels
- data_format information is a layer attribute

**Comment by fchollet** [comment]
This bug has been fixed in: https://github.com/fchollet/keras/pull/6658

Thanks!
